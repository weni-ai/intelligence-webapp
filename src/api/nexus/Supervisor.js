import nexusRequest from '../nexusaiRequest';
import { ConversationAdapter } from '../adapters/supervisor/conversation';

export const Supervisor = {
  conversations: {
    async list(filterData) {
      const {
        projectUuid,
        signal,
        hideGenericErrorAlert = false,
        filters = {},
      } = filterData;

      const params = ConversationAdapter.toApi({ ...filters });

      const config = { signal, hideGenericErrorAlert };

      const { data } = await nexusRequest.$http.get(
        `/api/${projectUuid}/supervisor/?${new URLSearchParams(params)}`,
        config,
      );

      return ConversationAdapter.fromApi(data);
    },

    async getById({ projectUuid, start, end, urn, next }) {
      const params = {
        start,
        end,
        contact_urn: urn,
      };

      let url = `/api/${projectUuid}/conversations/?${new URLSearchParams(params)}`;

      if (next) {
        url = next.slice(next.indexOf('/api'));
      }

      const { data } = await nexusRequest.$http.get(url);

      return data;
    },

    async getTopics({ projectUuid }) {
      const { data } = await nexusRequest.$http.get(
        `/api/${projectUuid}/topics/`,
      );
      return data;
    },

    async export({ projectUuid, token }) {
      const { data } = await nexusRequest.$http.post('/api/reports', {
        auth_token: token,
        project_uuid: projectUuid,
      });
      return data;
    },

    async getExportEmails({ projectUuid }) {
      const { data } = await nexusRequest.$http.get(
        `/api/reports?project_uuid=${projectUuid}`,
      );
      return data;
    },
  },
};
