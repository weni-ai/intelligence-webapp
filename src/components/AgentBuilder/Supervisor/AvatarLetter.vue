<template>
  <section class="avatar-letter">
    <p class="avatar-letter__text">{{ displayText }}</p>
  </section>
</template>

<script setup>
import { computed } from 'vue';

const props = defineProps({
  text: {
    type: String,
    default: '',
  },
  fallback: {
    type: String,
    default: '-',
  },
});

const displayText = computed(() => {
  return props.text?.[0] || props.fallback;
});
</script>

<style lang="scss" scoped>
.avatar-letter {
  $size: 35px;

  min-width: $size;
  min-height: $size;
  max-width: $size;
  max-height: $size;

  border-radius: 50%;
  background-color: $unnnic-color-weni-100;

  display: flex;
  align-items: center;
  justify-content: center;

  &__text {
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-bold;
    color: $unnnic-color-weni-600;
  }
}
</style>
