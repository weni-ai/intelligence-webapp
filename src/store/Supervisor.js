import { defineStore } from 'pinia';
import { computed, reactive } from 'vue';
import globalStore from '.';

import nexusaiAPI from '@/api/nexusaiAPI';

import { PerformanceAdapter } from '@/api/adapters/supervisor/performance';

export const useSupervisorStore = defineStore('Supervisor', () => {
  const projectUuid = computed(() => globalStore.state.Auth.connectProjectUuid);
  const supervisorApi = nexusaiAPI.agent_builder.supervisor;

  const forwardStats = reactive({
    status: null,
    data: {
      attendedByAgent: 0,
      forwardedHumanSupport: 0,
    },
  });

  const conversations = reactive({
    status: null,
    data: [],
  });

  async function loadConversations() {
    conversations.status = 'loading';
    try {
      const response = await supervisorApi.conversations.list({
        projectUuid: projectUuid.value,
        start: '2025-01-01',
        end: '2025-05-01',
      });

      conversations.status = 'complete';
      conversations.data = response;
    } catch (error) {
      conversations.status = 'error';
    }
  }

  async function loadForwardStats() {
    forwardStats.status = 'loading';
    try {
      const response = await supervisorApi.conversations.forwardStats({
        projectUuid: projectUuid.value,
        start: '2025-01-01',
        end: '2025-05-01',
      });

      const adaptedData = PerformanceAdapter.fromApi(response);
      const total =
        adaptedData.attendedByAgent + adaptedData.forwardedHumanSupport;

      const calculatePercentage = (value) =>
        total === 0 ? 0 : Math.round((value / total) * 100);

      forwardStats.data = {
        attendedByAgent: calculatePercentage(adaptedData.attendedByAgent),
        forwardedHumanSupport: calculatePercentage(
          adaptedData.forwardedHumanSupport,
        ),
      };

      forwardStats.status = 'complete';
    } catch (error) {
      forwardStats.status = 'error';
    }
  }

  return {
    forwardStats,
    loadForwardStats,
    conversations,
    loadConversations,
  };
});
