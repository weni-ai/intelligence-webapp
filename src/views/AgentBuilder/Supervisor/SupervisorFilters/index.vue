<template>
  <section class="supervisor-filters">
    <FilterText data-testid="filter-text" />

    <UnnnicButton
      data-testid="button-filter"
      iconLeft="filter_list"
      type="secondary"
      :text="filterButtonText"
      @click="openFilterDrawer"
    />

    <UnnnicDrawer
      data-testid="drawer-filter"
      :modelValue="isFilterDrawerOpen"
      class="supervisor-filters__drawer"
      :title="
        $t('agent_builder.supervisor.filters.filter_conversations_drawer')
      "
      :primaryButtonText="$t('agent_builder.supervisor.filters.apply_filters')"
      :disabledPrimaryButton="filterDrawerApplyButtonDisabled"
      :secondaryButtonText="
        $t('agent_builder.supervisor.filters.clear_filters')
      "
      :disabledSecondaryButton="filterDrawerClearButtonDisabled"
      @close="closeFilterDrawer"
      @primary-button-click="applyFilters"
      @secondary-button-click="clearFilters"
    >
      <template #content>
        <FilterDate data-testid="filter-date" />
        <FilterStatus data-testid="filter-status" />
        <FilterCsat data-testid="filter-csat" />
        <FilterTopics data-testid="filter-topics" />
      </template>
    </UnnnicDrawer>
  </section>
</template>

<script setup>
import { ref, computed, onMounted } from 'vue';
import { isEqual } from 'lodash';

import i18n from '@/utils/plugins/i18n';

import { useSupervisorStore } from '@/store/Supervisor';

import FilterText from './FilterText.vue';
import FilterDate from './FilterDate.vue';
import FilterStatus from './FilterStatus.vue';
import FilterCsat from './FilterCsat.vue';
import FilterTopics from './FilterTopics.vue';

const supervisorStore = useSupervisorStore();

const isFilterDrawerOpen = ref(false);
const filterDrawerApplyButtonDisabled = computed(() =>
  isEqual(supervisorStore.temporaryFilters, supervisorStore.filters),
);

const filterDrawerClearButtonDisabled = computed(() =>
  isEqual(supervisorStore.temporaryFilters, supervisorStore.defaultFilters),
);

const countAppliedFilters = computed(() => {
  const filtersToCount = ['status', 'csat', 'topics'];

  return filtersToCount.reduce((total, filter) => {
    return total + supervisorStore.filters[filter].length;
  }, 0);
});

const filterButtonText = computed(() => {
  const { t, tc } = i18n.global;
  const count = countAppliedFilters.value;

  const text = t('agent_builder.supervisor.filters.filter_conversations');
  const countText = tc(
    'agent_builder.supervisor.filters.count_applied_filters',
    count,
    {
      count,
    },
  );

  return count > 0 ? `${text} ${countText}` : text;
});

function openFilterDrawer() {
  isFilterDrawerOpen.value = true;
}

function closeFilterDrawer() {
  isFilterDrawerOpen.value = false;
}

function clearFilters() {
  supervisorStore.resetFilters();
  closeFilterDrawer();
}

function applyFilters() {
  supervisorStore.updateFilters();
  closeFilterDrawer();
}

onMounted(async () => {
  await supervisorStore.getTopics();
});
</script>

<style scoped lang="scss">
.supervisor-filters {
  display: grid;
  grid-template-columns: 8fr 4fr;
  gap: $unnnic-spacing-sm;

  &__drawer {
    :deep(.unnnic-drawer__container .unnnic-drawer__content) {
      overflow: visible;

      display: flex;
      flex-direction: column;
      gap: $unnnic-spacing-sm;
    }
  }
}
</style>
