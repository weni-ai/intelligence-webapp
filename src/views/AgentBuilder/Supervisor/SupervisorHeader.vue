<template>
  <AgentBuilderHeader :withDivider="false">
    <template #actions>
      <UnnnicInputDatePicker
        v-model="dateFilter"
        position="right"
        class="supervisor-header__date-picker"
        :maxDate="today"
      />
    </template>
  </AgentBuilderHeader>
</template>

<script setup>
import AgentBuilderHeader from '@/components/AgentBuilder/Header.vue';

import { useSupervisorStore } from '@/store/Supervisor';

import { format, subDays } from 'date-fns';
import { ref, watch } from 'vue';

const supervisorStore = useSupervisorStore();

const last7Days = format(subDays(new Date(), 7), 'yyyy-MM-dd');
const today = format(new Date(), 'yyyy-MM-dd');

const dateFilter = ref({
  start: last7Days,
  end: today,
});

watch(
  () => dateFilter.value,
  () => {
    supervisorStore.filters.start = dateFilter.value.start;
    supervisorStore.filters.end = dateFilter.value.end;
  },
  { immediate: true },
);
</script>

<style lang="scss" scoped>
.supervisor-header {
  &__date-picker {
    :deep(.input) {
      width: 100%;
    }
  }
}
</style>
