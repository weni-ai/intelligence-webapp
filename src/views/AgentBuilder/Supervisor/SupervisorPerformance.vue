<template>
  <section class="supervisor-performance">
    <PerformanceCard
      v-for="stat in performanceStats"
      :key="stat.title"
      data-test="monitoring-performance-card"
      :title="stat.title"
      :tooltip="stat.tooltip"
      :value="stat.value"
      :scheme="stat.scheme"
      :isLoading="supervisorStore.forwardStats.status === 'loading'"
    />
  </section>
</template>

<script setup>
import { onMounted, computed } from 'vue';

import { useSupervisorStore } from '@/store/Supervisor';

import PerformanceCard from '@/components/Brain/Monitoring/PerformanceCard.vue';
import i18n from '@/utils/plugins/i18n';

const supervisorStore = useSupervisorStore();
const t = (key) => i18n.global.t(key);

const performanceStats = computed(() => {
  const stats = supervisorStore.forwardStats.data;

  return [
    {
      title: t('agent_builder.supervisor.attended_by_agent.title'),
      tooltip: t('agent_builder.supervisor.attended_by_agent.tooltip'),
      value: stats.attendedByAgent,
      scheme: 'green',
    },
    {
      title: t('agent_builder.supervisor.forwarded_human_support.title'),
      tooltip: t('agent_builder.supervisor.forwarded_human_support.tooltip'),
      value: stats.forwardedHumanSupport,
      scheme: 'blue',
    },
  ];
});

onMounted(() => {
  supervisorStore.loadForwardStats();
});
</script>

<style lang="scss" scoped>
.supervisor-performance {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  gap: $unnnic-spacing-sm;
}
</style>
