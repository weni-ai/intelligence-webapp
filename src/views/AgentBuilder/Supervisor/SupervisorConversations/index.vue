<template>
  <section
    :class="['conversations', { 'conversations--empty': !hasConversations }]"
  >
    <SupervisorFilters data-testid="supervisor-filters" />

    <ConversationsTable
      ref="conversationsTable"
      data-testid="conversations-table"
    />
  </section>
</template>

<script setup>
import { ref, defineExpose, computed } from 'vue';

import SupervisorFilters from '../SupervisorFilters/index.vue';
import ConversationsTable from './ConversationsTable/index.vue';
import { useSupervisorStore } from '@/store/Supervisor';

const supervisorStore = useSupervisorStore();

const hasConversations = computed(
  () => supervisorStore.conversations.data.results.length > 0,
);

const conversationsTable = ref(null);

defineExpose({
  loadMoreConversations: () => conversationsTable.value.loadMoreConversations(),
});
</script>

<style scoped lang="scss">
.conversations {
  display: grid;
  grid-template-rows: auto 1fr;
  gap: $unnnic-spacing-sm;
  align-items: start;

  &--empty {
    height: 100%;
  }
}
</style>
