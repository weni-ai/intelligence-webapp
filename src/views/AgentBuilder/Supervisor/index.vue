<template>
  <section
    :class="[
      'supervisor',
      { 'supervisor--with-conversation': selectedConversation },
    ]"
  >
    <section
      class="supervisor__content"
      @scroll="loadConversations"
    >
      <SupervisorHeader
        class="supervisor__header"
        data-testid="header"
      />
      <SupervisorConversations
        ref="supervisorConversations"
        class="supervisor__conversations"
        data-testid="supervisor-conversations"
      />
    </section>
    <Conversation
      v-if="selectedConversation"
      class="supervisor__conversation"
      data-testid="supervisor-conversation"
    />
  </section>
</template>

<script setup>
import { computed, onBeforeMount, watch, ref } from 'vue';
import { useRouter, useRoute } from 'vue-router';

import SupervisorHeader from './SupervisorHeader.vue';
import SupervisorConversations from './SupervisorConversations/index.vue';
import Conversation from './SupervisorConversations/Conversation/index.vue';

import { useSupervisorStore } from '@/store/Supervisor';
import { cleanParams } from '@/utils/http';

const supervisorStore = useSupervisorStore();
const router = useRouter();
const route = useRoute();

const supervisorConversations = ref(null);

const selectedConversation = computed(() => {
  return supervisorStore.selectedConversation;
});

function updateQuery(filters = supervisorStore.filters) {
  const cleanedFilters = cleanParams(filters);
  router.replace({
    query: {
      ...cleanedFilters,
    },
  });
}

function loadConversations(event) {
  const { next } = supervisorStore.conversations.data;

  if (
    !next ||
    ['loading', 'error'].includes(supervisorStore.conversations.status)
  ) {
    return;
  }

  const { scrollTop, clientHeight, scrollHeight } = event.target;

  const safeDistance = 10;
  const isInScrollBottom =
    scrollTop + clientHeight + safeDistance >= scrollHeight;

  const shouldLoadMore = isInScrollBottom;

  if (shouldLoadMore) {
    supervisorConversations.value.loadMoreConversations();
  }
}

watch(
  () => supervisorStore.filters,
  () => {
    updateQuery();
  },
  { deep: true },
);

watch(
  () => supervisorStore.queryConversationUuid,
  (conversationUuid) => {
    updateQuery({
      ...route.query,
      conversationUuid,
    });
  },
);

onBeforeMount(async () => {
  updateQuery();

  const { selectedConversation, queryConversationUuid } = supervisorStore;

  if (queryConversationUuid && !selectedConversation) {
    supervisorStore.selectConversation(queryConversationUuid);
  }
});
</script>

<style lang="scss" scoped>
.supervisor {
  margin: -$unnnic-spacing-sm;

  min-height: 100%;

  display: grid;

  overflow: hidden;

  &--with-conversation {
    grid-template-columns: 7fr 5fr;
  }

  &__content {
    display: flex;
    flex-direction: column;

    $scroll-margin: calc($unnnic-spacing-nano / 2 + $unnnic-spacing-nano);

    overflow-y: auto;

    margin: $unnnic-spacing-sm 0;
    padding-right: $scroll-margin;
    margin-right: $scroll-margin;
  }

  :deep(.supervisor__header) {
    padding: 0 $unnnic-spacing-sm $unnnic-spacing-md;
  }

  &__conversations {
    padding-left: $unnnic-spacing-sm;

    & > * {
      margin-right: $unnnic-spacing-sm;
    }
  }
}
</style>
