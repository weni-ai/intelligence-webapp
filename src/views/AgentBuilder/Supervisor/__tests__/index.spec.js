import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { createPinia, setActivePinia } from 'pinia';
import { useRouter } from 'vue-router';

import Supervisor from '@/views/AgentBuilder/Supervisor/index.vue';
import { useSupervisorStore } from '@/store/Supervisor';

vi.mock('vue-router', () => ({
  useRoute: vi.fn(() => ({
    query: {
      started_day: '2024-01-01',
      ended_day: '2024-01-31',
    },
  })),
  useRouter: vi.fn(() => ({
    replace: vi.fn(),
  })),
}));

describe('Supervisor view', () => {
  let wrapper;
  let supervisorStore;
  let router;

  beforeEach(() => {
    const pinia = createPinia();
    setActivePinia(pinia);

    supervisorStore = useSupervisorStore();

    router = useRouter();

    wrapper = shallowMount(Supervisor, {
      global: {
        plugins: [pinia],
      },
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders SupervisorHeader component', () => {
    expect(wrapper.findComponent('[data-testid="header"]').exists()).toBe(true);
  });

  it('renders SupervisorConversations component', () => {
    expect(
      wrapper
        .findComponent('[data-testid="supervisor-conversations"]')
        .exists(),
    ).toBe(true);
  });

  it('renders Conversation component when selectedConversation is present', async () => {
    expect(
      wrapper.findComponent('[data-testid="supervisor-conversation"]').exists(),
    ).toBe(false);

    supervisorStore.selectedConversation = {
      id: 1,
      title: 'Test Conversation',
    };

    await wrapper.vm.$nextTick();

    expect(wrapper.classes()).toContain('supervisor--with-conversation');
    expect(
      wrapper.findComponent('[data-testid="supervisor-conversation"]').exists(),
    ).toBe(true);
  });
});
