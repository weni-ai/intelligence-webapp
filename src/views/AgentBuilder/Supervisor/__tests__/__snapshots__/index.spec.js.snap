// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Supervisor view > matches snapshot 1`] = `
"<section data-v-50a82958="" class="supervisor">
  <section data-v-50a82958="" class="supervisor__content">
    <supervisor-header-stub data-v-50a82958="" class="supervisor__header" data-testid="header"></supervisor-header-stub>
    <supervisor-conversations-stub data-v-50a82958="" class="supervisor__conversations" data-testid="supervisor-conversations"></supervisor-conversations-stub>
  </section>
  <!--v-if-->
</section>"
`;
